File: nomad-samples/python/basic-live-channel-1.1/live_input/wait_live_input_status.py
from live_input.get_live_input_status import *
from live_input.get_live_input_status_message import *

from libraries import time

'''
 * Wait for a Live Input to transition to specified status
 *
 * @param {string} AUTH_TOKEN                    | Authorization AUTH_TOKEN
 * @param {string} INPUT_ID                      | The Live Input ID for which to get the status
 * @param {LIVE_INPUT_STATUSES} STATUS_TO_WAIT_FOR   | The live Live Input status to wait for
 * @param {number} TIMEOUT                      | Timeout in seconds
 * @param {number} POLL_INTERVAL                 | Poll interval in seconds
'''
async def wait_for_live_input_status(AUTH_TOKEN, INPUT_ID, STATUS_TO_WAIT_FOR, TIMEOUT = 30, POLL_INTERVAL = 2):
    # Check for valid parameters
    if (not AUTH_TOKEN or not INPUT_ID or not STATUS_TO_WAIT_FOR):
        raise Exception("Wait for Live Input Status: Invalid API call")


    # Set the starting time
    STARTING_TIME = time.time()

    # Elapsed time in seconds
    ELAPSED_TIME = 0

    while (ELAPSED_TIME < TIMEOUT):
        # Get the Live Input status
        INPUT_STATUS = await get_live_input_status(AUTH_TOKEN, INPUT_ID)

        # If Live Input is in STATUS_TO_WAIT_FOR return
        if (INPUT_STATUS == STATUS_TO_WAIT_FOR):
            # Give feedback to the console
            print("Live Input " + str(INPUT_ID) + " transitioned to status " + STATUS_TO_WAIT_FOR)
            return


        # Give feedback to the console
        print("Live Input [" + INPUT_ID + "] is in status [" + INPUT_STATUS + "]")

        # Check for Error status
        if (INPUT_STATUS == "Error"):
            # Get the error message
            INPUT_STATUS_MESSAGE = await get_live_input_status_message(AUTH_TOKEN, INPUT_ID)

            # Can't continue if Live Input is in error status
            raise Exception("Live Input " + str(INPUT_ID) + " is in Error status: " + INPUT_STATUS_MESSAGE)


        # Calculate elapsed time in seconds
        ELAPSED_TIME = (time.time() - STARTING_TIME)

        # Give feedback to the console
        print("Waiting for Live Input [" + str(INPUT_ID) + "] to transition to status [" + STATUS_TO_WAIT_FOR + "]... [" + str(round(ELAPSED_TIME)) + "] timeout:" + str(TIMEOUT) + "]")

        # Check for TIMEOUT
        if (ELAPSED_TIME > TIMEOUT):
            break


        # Wait poll interval
        time.sleep(POLL_INTERVAL)


    raise Exception("Waiting for Live Input [" + INPUT_ID + "] to transition to status [" + STATUS_TO_WAIT_FOR + "] timed out")

File: nomad-samples/admin/js/advanced-live-channel-1.1/live-input/wait-live-input-status.js
import getLiveInputStatus from "./get-live-input-status.js";
import getLiveInputStatusMessage from "./get-live-input-status-message.js";
import sleep from "../helpers/sleep.js";

/**
 * Wait for a Live Input to transition to specified status
 *
 * @param {string} authToken                    | Authorization authToken
 * @param {string} inputId                      | The Live Input ID for which to get the status
 * @param {liveInputStatuses} statusToWaitFor   | The live Live Input status to wait for
 * @param {number} timeout                      | Timeout in seconds
 * @param {number} pollInterval                 | Poll interval in seconds
 */
export default async function waitForLiveInputStatus(authToken, inputId, statusToWaitFor, timeout = 30, pollInterval = 2) {
    // Check for valid parameters
    if (!authToken || !inputId || !statusToWaitFor) {
        throw new Error("Wait for Live Input Status: Invalid API call");
    }

    // Set the starting time
    const startingTime = new Date().getTime();

    // Elapsed time in seconds
    let elapsedTime = 0;

    while (elapsedTime < timeout) {
        // Get the Live Input status
        const inputStatus = await getLiveInputStatus(authToken, inputId);

        // If Live Input is in statusToWaitFor return
        if (inputStatus === statusToWaitFor) {
            // Give feedback to the console
            console.log(`Live Input ${inputId} transitioned to status ${statusToWaitFor}`);
            return;
        }

        // Give feedback to the console
        console.log(`Live Input [${inputId}] is in status [${inputStatus}]`);

        // Check for Error status
        if (inputStatus === "Error") {
            // Get the error message
            const inputStatusMessage = await getLiveInputStatusMessage(authToken, inputId);

            // Can't continue if Live Input is in error status
            throw new Error(`Live Input ${inputId} is in Error status: ${inputStatusMessage}`);
        }

        // Calculate elapsed time in seconds
        elapsedTime = (new Date().getTime() - startingTime) / 1000;

        // Give feedback to the console
        console.log(`Waiting for Live Input [${inputId}] to transition to status [${statusToWaitFor}]... [${Math.round(elapsedTime)}/${timeout}]`);

        // Check for timeout
        if (elapsedTime > timeout) {
            break;
        }

        // Wait poll interval
        await sleep(pollInterval);
    }

    throw new Error(`Waiting for Live Input [${inputId}] to transition to status [${statusToWaitFor}] timed out`);
}

/**

 */
